pallets/salp/src/lib.rs

```diff
@@ -306,112 +429,139 @@ pub mod pallet {
 		/// slot is unable to be purchased and the timeout expires.
 		#[pallet::weight(0)]
 		pub(super) fn contribute(
-			_origin: OriginFor<T>,
+			origin: OriginFor<T>,
 			#[pallet::compact] index: ParaId,
-			contributor: AccountId<T>,
 			#[pallet::compact] value: BalanceOf<T>,
 		) -> DispatchResultWithPostInfo {
-			let who = contributor.clone();//ensure_signed(origin)?;
-
-			ensure!(value >= T::MinContribution::get(), Error::<T>::ContributionTooSmall);
-			let mut fund = Self::funds(index).ok_or(Error::<T>::InvalidParaId)?;
-			ensure!(fund.status == FundStatus::Ongoing, Error::<T>::InvalidFundStatus);
-			fund.raised  = fund.raised.checked_add(&value).ok_or(Error::<T>::Overflow)?;
-			ensure!(fund.raised <= fund.cap, Error::<T>::CapExceeded);
-
-			let old_balance = Self::contribution_get(fund.trie_index, &who);
+			let who = ensure_signed(origin.clone())?;
 
-			let balance = old_balance.saturating_add(value);
-			Self::contribution_put(fund.trie_index, &who, &balance);
+			ensure!(
+				value >= T::MinContribution::get(),
+				Error::<T>::ContributionTooSmall
+			);
 
-			// TODO deposit KSM/DOT to fund_account
-			// issue vsToken/vsBond to sender
-			// call XcmHandler with cross-chain transaction to contribute KSM/DOT on relay chain
+			let fund = Self::funds(index).ok_or(Error::<T>::InvalidParaId)?;
+			ensure!(
+				fund.status == FundStatus::Ongoing,
+				Error::<T>::InvalidFundStatus
+			);
+			fund.raised
+				.checked_add(&value)
+				.ok_or(Error::<T>::Overflow)?;
+			ensure!(fund.raised <= fund.cap, Error::<T>::CapExceeded);
 
-			Funds::<T>::insert(index, Some(fund));
+			Self::xcm_ump_contribute(origin, index, value).map_err(|_e| Error::<T>::XcmFailed)?;
 
-			Self::deposit_event(Event::Contributed(who, index, value));
+			Self::deposit_event(Event::Contributing(who, index, value));
 
 			Ok(().into())
 		}
 
-		/// Withdraw full balance of a contributor.
-		///
-		/// Origin must be signed.
-		///
-		/// The fund must be either in, or ready for, retirement. For a fund to be *in* retirement, then the retirement
-		/// flag must be set. For a fund to be ready for retirement, then:
-		/// - it must not already be in retirement;
-		/// - the amount of raised funds must be bigger than the _free_ balance of the account;
-		/// - and either:
-		///   - the block number must be at least `end`; or
-		///   - the current lease period must be greater than the fund's `last_slot`.
-		///
-		/// In this case, the fund's retirement flag is set and its `end` is reset to the current block
-		/// number.
-		///
-		/// - `who`: The account whose contribution should be withdrawn.
+		/// Withdraw full balance of the parachain.
 		/// - `index`: The parachain to whose crowdloan the contribution was made.
-		// #[weight = T::WeightInfo::withdraw()]
 		#[pallet::weight(0)]
 		pub(super) fn withdraw(
 			origin: OriginFor<T>,
-			who: AccountId<T>,
 			#[pallet::compact] index: ParaId,
```

Comment:
the withdraw here is just to withdraw all raised from crowdloan account to parachain account with xcm transact